typing/typecore.ml

```diff
@@ -1334,27 +1352,34 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
       let opath, record_ty =
         try
           let (p0, p,_) = extract_concrete_record !env expected_ty in
-          Some (p0, p, true), expected_ty
+          begin_def ();
+          let ty = instance expected_ty in
+          end_def ();
+          generalize_structure ty;
+          Some (p0, p, true), ty
         with Not_found -> None, newvar ()
       in
       let type_label_pat (label_lid, label, sarg) k =
+        begin_def ();
         let (_, ty_arg, ty_res) = instance_label false label in
         begin try
-          unify_pat_types loc !env ty_res record_ty
+          unify_pat_types loc !env ty_res (instance record_ty)
         with Error(_loc, _env, Pattern_type_clash(trace)) ->
           raise(Error(label_lid.loc, !env,
                       Label_mismatch(label_lid.txt, trace)))
         end;
+        end_def ();
+        generalize_structure ty_res;
+        generalize_structure ty_arg;
         type_pat sarg ty_arg (fun arg ->
           k (label_lid, label, arg))
       in
       let k' k lbl_pat_list =
         check_recordpat_labels loc lbl_pat_list closed;
-        unify_pat_types loc !env record_ty expected_ty;
-        rp k {
+        rup k {
         pat_desc = Tpat_record (lbl_pat_list, closed);
         pat_loc = loc; pat_extra=[];
-        pat_type = expected_ty;
+        pat_type = record_ty;
         pat_attributes = sp.ppat_attributes;
```

Comment:
Shouldn't it be `instance record_ty` ? And why call `rup` rather than `rp` (`record_ty` is already an instance of `expected_ty`) ?

Reply:
It should be `instance record_ty` you are correct.
However `record_ty` is not necessarily an instance of `expected_ty`, in fact in most cases it won't be: it will be only when `extract_concrete_record` succeeds.

Reply:
Note: I pushed a commit adding the instance taking.